boa_engine/src/nonmaxu32.rs

```diff
@@ -0,0 +1,180 @@
+        "unable to parse integer".fmt(fmt)
+    }
+}
+
+impl From<core::num::ParseIntError> for ParseIntError {
+    fn from(_: core::num::ParseIntError) -> Self {
+        Self(())
+    }
+}
+
+impl NonMaxU32 {
+    /// Creates a non-u32-max without checking the value.
+    ///
+    /// # Safety
+    ///
+    /// The value must not be `u32::MAX`.
+    #[inline]
+    pub const unsafe fn new_unchecked(n: u32) -> Self {
+        // SAFETY: this is guaranteed to be safe by the caller.
+        Self(n)
+    }
+
+    /// Creates a non-u32-max if the given value is not `u32::MAX`.
+    #[inline]
+    pub const fn new(n: u32) -> Option<Self> {
+        if n == u32::MAX {
+            None
+        } else {
+            // SAFETY: we just checked that there's no `u32::MAX`
+            Some(Self(n))
+        }
+    }
+
+    /// Returns the value as a primitive type.
+    #[inline]
+    pub const fn get(self) -> u32 {
+        self.0
+    }
+}
+
+impl TryFrom<u32> for NonMaxU32 {
+    type Error = TryFromIntError;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        Self::new(value).ok_or(TryFromIntError(()))
+    }
+}
+
+impl From<NonMaxU32> for u32 {
+    /// Converts a `NonMaxU32` into an `u32`
+    fn from(nonzero: NonMaxU32) -> Self {
+        nonzero.0
+    }
+}
+
+impl core::str::FromStr for NonMaxU32 {
+    type Err = ParseIntError;
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        Self::new(u32::from_str(value)?).ok_or(ParseIntError(()))
+    }
+}
+
+impl BitOr for NonMaxU32 {
+    type Output = Self;
+    #[inline]
+    fn bitor(self, rhs: Self) -> Self::Output {
+        // Safety: since `self` and `rhs` are both nonzero, the
+        // result of the bitwise-or will be nonzero.
+        unsafe { NonMaxU32::new_unchecked(self.get() | rhs.get()) }
+    }
+}
+
+impl BitOr<u32> for NonMaxU32 {
+    type Output = Self;
+
+    #[inline]
+    fn bitor(self, rhs: u32) -> Self::Output {
+        // Safety: since `self` is nonzero, the result of the
+        // bitwise-or will be nonzero regardless of the value of
+        // `rhs`.
+        unsafe { NonMaxU32::new_unchecked(self.get() | rhs) }
+    }
+}
+
+impl BitOr<NonMaxU32> for u32 {
+    type Output = NonMaxU32;
+
+    #[inline]
+    fn bitor(self, rhs: NonMaxU32) -> Self::Output {
+        // Safety: since `rhs` is nonzero, the result of the
+        // bitwise-or will be nonzero regardless of the value of
+        // `self`.
+        unsafe { NonMaxU32::new_unchecked(self | rhs.get()) }
+    }
+}
+
+impl BitOrAssign for NonMaxU32 {
+    #[inline]
+    fn bitor_assign(&mut self, rhs: Self) {
+        *self = *self | rhs;
+    }
+}
+
+impl BitOrAssign<u32> for NonMaxU32 {
+    #[inline]
+    fn bitor_assign(&mut self, rhs: u32) {
+        *self = *self | rhs;
+    }
+}
+
+impl fmt::Debug for NonMaxU32 {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.get().fmt(f)
+    }
+}
```

Comment:
This might be a good implementation for `Display`, but I would say the `Debug` implementation shouldn't hide the internal structure. I'm open to suggestions, though.

Reply:
I tried to follow `NonZeroU32` debug though, it sounds fine to me to format it like `NonMaxU32 { inner: 42 }`.

Reply:
Agree with Razican